File: services/aas-verifier-py/src/aas_verifier/schema.py
"""Step 1 of the AAS v1 verifier — pure schema check.

Implements the well-formed-per-type table from
``docs/specs/aas-v1.md`` §4. Returns the first failure reason
encountered, or ``None`` if the attestation passes.
"""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any, Optional

import base58


U64_MAX = (1 << 64) - 1

# Anchor pattern: YYYY-MM-DDTHH:MM:SS{Z|+00:00}. No fractional seconds
# (per spec §3 — on-chain source is i64 unix seconds).
_RFC3339_NO_FRACTION = re.compile(
    r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(Z|\+00:00)$"
)
_DECIMAL_U64 = re.compile(r"^[0-9]+$")
_HEX32 = re.compile(r"^[0-9a-f]{64}$")


def check_schema(input_value: Any) -> Optional[str]:
    """Return the first ``schema_invalid:<field>`` reason, or ``None``."""
    if not isinstance(input_value, dict):
        return "schema_invalid:<root>"

    a = input_value

    if a.get("version") != "aas/v1":
        return "schema_invalid:version"

    if not _is_pubkey(a.get("program_id")):
        return "schema_invalid:program_id"
    if not _is_pubkey(a.get("account")):
        return "schema_invalid:account"
    if not _is_pubkey(a.get("client")):
        return "schema_invalid:client"
    if not _is_pubkey(a.get("agent")):
        return "schema_invalid:agent"

    oracle_feed = a.get("oracle_feed", "missing")
    if oracle_feed != "missing" and oracle_feed is not None and not _is_pubkey(oracle_feed):
        return "schema_invalid:oracle_feed"
    if oracle_feed == "missing":
        return "schema_invalid:oracle_feed"

    if a.get("network") not in ("mainnet", "devnet"):
        return "schema_invalid:network"

    if not _is_u8(a.get("platform")):
        return "schema_invalid:platform"

    if not _is_decimal_u64(a.get("task_id")):
        return "schema_invalid:task_id"
    if not _is_decimal_u64(a.get("composite_score")):
        return "schema_invalid:composite_score"
    if not _is_decimal_u64(a.get("score_max")):
        return "schema_invalid:score_max"

    if not _is_hex32(a.get("verification_hash")):
        return "schema_invalid:verification_hash"
    if not _is_hex32(a.get("content_hash")):
        return "schema_invalid:content_hash"
    if not _is_hex32(a.get("content_id_hash")):
        return "schema_invalid:content_id_hash"

    if not _is_enum_string(a.get("state")):
        return "schema_invalid:state"
    if not _is_enum_string(a.get("account_kind")):
        return "schema_invalid:account_kind"

    if not _is_rfc3339_no_fraction(a.get("verified_at")):
        return "schema_invalid:verified_at"

    extensions = a.get("extensions")
    if extensions is not None and not isinstance(extensions, dict):
        return "schema_invalid:extensions"

    return None


def _is_pubkey(v: Any) -> bool:
    """Base58, decodes to exactly 32 bytes."""
    if not isinstance(v, str) or not v:
        return False
    try:
        return len(base58.b58decode(v)) == 32
    except Exception:
        return False


def _is_u8(v: Any) -> bool:
    return isinstance(v, int) and not isinstance(v, bool) and 0 <= v <= 255


def _is_decimal_u64(v: Any) -> bool:
    if not isinstance(v, str) or not _DECIMAL_U64.fullmatch(v):
        return False
    if len(v) > 1 and v.startswith("0"):
        return False
    try:
        n = int(v)
    except ValueError:
        return False
    return 0 <= n <= U64_MAX


def _is_hex32(v: Any) -> bool:
    return isinstance(v, str) and bool(_HEX32.fullmatch(v))


def _is_enum_string(v: Any) -> bool:
    return isinstance(v, str) and len(v) > 0 and not any(ch.isspace() for ch in v)


def _is_rfc3339_no_fraction(v: Any) -> bool:
    if not isinstance(v, str):
        return False
    if not _RFC3339_NO_FRACTION.fullmatch(v):
        return False
    # Round-trip through datetime to confirm the components describe a
    # real moment (rejects e.g. "2026-13-32T25:61:99Z").
    s = v.replace("Z", "+00:00")
    try:
        datetime.fromisoformat(s)
    except ValueError:
        return False
    return True


def parse_verified_at_seconds(v: str) -> int:
    """Parse the wire ``verified_at`` to unix seconds. Caller is
    responsible for having already passed schema check."""
    s = v.replace("Z", "+00:00")
    dt = datetime.fromisoformat(s).astimezone(timezone.utc)
    return int(dt.timestamp())

File: sdk/aas-verifier-ts/src/schema.ts
import { PublicKey } from "@solana/web3.js";
import type { AasV1Attestation, FailureReason } from "./types.js";

/**
 * Step 1 — schema check. Implements the well-formed-per-type table
 * from `docs/specs/aas-v1.md` §4. Returns the first failure reason
 * encountered, or null if the schema check passes.
 *
 * Pure — no RPC, no I/O.
 */
export function checkSchema(input: unknown): FailureReason | null {
  if (typeof input !== "object" || input === null || Array.isArray(input)) {
    return "schema_invalid:<root>";
  }
  const a = input as Record<string, unknown>;

  if (a.version !== "aas/v1") return "schema_invalid:version";
  if (!isPubkey(a.program_id)) return "schema_invalid:program_id";
  if (!isPubkey(a.account)) return "schema_invalid:account";
  if (!isPubkey(a.client)) return "schema_invalid:client";
  if (!isPubkey(a.agent)) return "schema_invalid:agent";
  if (a.oracle_feed !== null && !isPubkey(a.oracle_feed)) {
    return "schema_invalid:oracle_feed";
  }

  if (a.network !== "mainnet" && a.network !== "devnet") {
    return "schema_invalid:network";
  }

  if (!isU8Number(a.platform)) return "schema_invalid:platform";

  if (!isDecimalU64(a.task_id)) return "schema_invalid:task_id";
  if (!isDecimalU64(a.composite_score)) return "schema_invalid:composite_score";
  if (!isDecimalU64(a.score_max)) return "schema_invalid:score_max";

  if (!isHex32(a.verification_hash)) return "schema_invalid:verification_hash";
  if (!isHex32(a.content_hash)) return "schema_invalid:content_hash";
  if (!isHex32(a.content_id_hash)) return "schema_invalid:content_id_hash";

  if (!isEnumString(a.state)) return "schema_invalid:state";
  if (!isEnumString(a.account_kind)) return "schema_invalid:account_kind";

  if (!isRfc3339NoFraction(a.verified_at)) {
    return "schema_invalid:verified_at";
  }

  if (a.extensions !== undefined && !isPlainObject(a.extensions)) {
    return "schema_invalid:extensions";
  }

  if (
    a.verifier_instructions !== undefined &&
    typeof a.verifier_instructions !== "string"
  ) {
    return "schema_invalid:verifier_instructions";
  }

  return null;
}

/** Base58, decodes to exactly 32 bytes. */
function isPubkey(v: unknown): boolean {
  if (typeof v !== "string" || v.length === 0) return false;
  try {
    const bytes = new PublicKey(v).toBytes();
    return bytes.length === 32;
  } catch {
    return false;
  }
}

/** JSON number in [0, 255]. */
function isU8Number(v: unknown): boolean {
  return typeof v === "number" && Number.isInteger(v) && v >= 0 && v <= 255;
}

/**
 * Decimal string matching ^[0-9]+$, no leading zeros except the literal
 * "0", fits in u64 (max 18446744073709551615 = 2^64 - 1).
 */
function isDecimalU64(v: unknown): boolean {
  if (typeof v !== "string") return false;
  if (!/^[0-9]+$/.test(v)) return false;
  if (v.length > 1 && v.startsWith("0")) return false;
  // Compare as BigInt to avoid Number precision loss.
  try {
    const n = BigInt(v);
    return n >= 0n && n <= U64_MAX;
  } catch {
    return false;
  }
}

const U64_MAX = (1n << 64n) - 1n;

/** Lowercase hex of length 64 (= 32 bytes). */
function isHex32(v: unknown): boolean {
  return typeof v === "string" && /^[0-9a-f]{64}$/.test(v);
}

/** Non-empty, no whitespace. Lowercase optional (verifiers tolerate). */
function isEnumString(v: unknown): boolean {
  return typeof v === "string" && v.length > 0 && !/\s/.test(v);
}

/**
 * RFC 3339 timestamp with NO fractional seconds. Either `Z` or `+00:00`
 * UTC offset (per spec §3 row for `verified_at`). Verifiers MUST accept
 * both forms.
 */
function isRfc3339NoFraction(v: unknown): boolean {
  if (typeof v !== "string") return false;
  // Anchor pattern: YYYY-MM-DDTHH:MM:SS{Z|+00:00}
  // No fractional seconds (".123" is rejected).
  const pattern =
    /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(Z|\+00:00)$/;
  if (!pattern.test(v)) return false;
  // Round-trip via Date to confirm the components describe a real
  // moment (rejects e.g. "2026-13-32T25:61:99Z").
  const ms = Date.parse(v);
  return !Number.isNaN(ms);
}

function isPlainObject(v: unknown): boolean {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/**
 * Once `checkSchema` passes, narrow the input to `AasV1Attestation`.
 * Pure runtime cast — no defensive checks (those already ran).
 */
export function asV1(a: unknown): AasV1Attestation {
  return a as AasV1Attestation;
}
